### PR TITLE
Fixes to reports.py -- don't crash on control characters, ensure directories earlier

### LINKED ISSUE #5884: Archaic control character 0x1f causes crash in linecount report
I have this file (foo.py):
```py
def foo():
    # ()
    return ""
```
The comment contains a weird control character hex 1f. Running mypy like this:
```
mypy --xml-report .xml-report foo.py --tb
```
causes this crash:
```
foo.py: error: INTERNAL ERROR -- please report a bug at https://github.com/python/mypy/issues version: 0.650+dev.94c54aa755f2d7826fed5e0e2276053ba49da8e6
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/Users/guido/src/server/.mypy/venv/lib/python3.6/site-packages/mypy/__main__.py", line 11, in <module>
    main(None)
  File "/Users/guido/src/server/.mypy/venv/lib/python3.6/site-packages/mypy/main.py", line 92, in main
    res = build.build(sources, options, None, flush_errors, fscache)
  File "/Users/guido/src/server/.mypy/venv/lib/python3.6/site-packages/mypy/build.py", line 155, in build
    result = _build(sources, options, alt_lib_path, flush_errors, fscache)
  File "/Users/guido/src/server/.mypy/venv/lib/python3.6/site-packages/mypy/build.py", line 204, in _build
    graph = dispatch(sources, manager)
  File "/Users/guido/src/server/.mypy/venv/lib/python3.6/site-packages/mypy/build.py", line 2172, in dispatch
    process_graph(graph, manager)
  File "/Users/guido/src/server/.mypy/venv/lib/python3.6/site-packages/mypy/build.py", line 2469, in process_graph
    process_stale_scc(graph, scc, manager)
  File "/Users/guido/src/server/.mypy/venv/lib/python3.6/site-packages/mypy/build.py", line 2605, in process_stale_scc
    graph[id].finish_passes()
  File "/Users/guido/src/server/.mypy/venv/lib/python3.6/site-packages/mypy/build.py", line 1767, in finish_passes
    manager.report_file(self.tree, self.type_map(), self.options)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/contextlib.py", line 99, in __exit__
    self.gen.throw(type, value, traceback)
  File "/Users/guido/src/server/.mypy/venv/lib/python3.6/site-packages/mypy/build.py", line 1519, in wrap_context
    yield
  File "/Users/guido/src/server/.mypy/venv/lib/python3.6/site-packages/mypy/build.py", line 1767, in finish_passes
    manager.report_file(self.tree, self.type_map(), self.options)
  File "/Users/guido/src/server/.mypy/venv/lib/python3.6/site-packages/mypy/build.py", line 643, in report_file
    self.reports.file(file, type_map, options)
  File "/Users/guido/src/server/.mypy/venv/lib/python3.6/site-packages/mypy/report.py", line 83, in file
    reporter.on_file(tree, type_map, options)
  File "/Users/guido/src/server/.mypy/venv/lib/python3.6/site-packages/mypy/report.py", line 450, in on_file
    any_info=self._get_any_info_for_line(visitor, lineno))
  File "src/lxml/etree.pyx", line 3112, in lxml.etree.SubElement
  File "src/lxml/apihelpers.pxi", line 199, in lxml.etree._makeSubElement
  File "src/lxml/apihelpers.pxi", line 194, in lxml.etree._makeSubElement
  File "src/lxml/apihelpers.pxi", line 297, in lxml.etree._initNodeAttributes
  File "src/lxml/apihelpers.pxi", line 311, in lxml.etree._addAttributeToNode
  File "src/lxml/apihelpers.pxi", line 1439, in lxml.etree._utf8
ValueError: All strings must be XML compatible: Unicode or ASCII, no NULL bytes or control characters
foo.py: : note: use --pdb to drop into pdb
Generated XML report: /Users/guido/src/server/.xml-report/index.xml
```

### PR DESCRIPTION
**Pull request by gvanrossum**
Fixes #5884 (both parts).

### CODE REVIEW

**Comment by gvanrossum** [comment]
I don't understand why the tests fail on Windows only, but clearly it's got something to do with my moving around the `ensure_dir_exists()` calls. Perhaps the temporary directory is relative on Windows but absolute on *nix?

**Comment by gvanrossum** [comment]
@msullivan This is now officially ready for review. I did expand a bit on the original issue (control characters) but I think the report generator and the test framework are a bit better now.

File: mypy/report.py [hunk 1]
```diff
@@ -418,6 +415,10 @@ def __init__(self, reports: Reports, output_dir: str) -> None:
         self.last_xml = None  # type: Optional[Any]
         self.files = []  # type: List[FileInfo]
 
+    # XML doesn't like control characters, but they are sometimes
+    # legal in source code (e.g. comments, string literals).
+    control_fixer = str.maketrans(''.join(chr(i) for i in range(32)), '?' * 32)
```
Comment (msullivan) [inline mypy/report.py]: I didn't know this function!

**Comment by gvanrossum** [comment]
> Why didn't the thing cause trouble before?

A comment containing an odd control character was checked into our S code base last week.

I fixed the test harness because I had a hell of a time debugging why the AppVeyor tests were failing (apparently you can't have a directory named `<memory>` on Windows, but the test harness was ignoring the traceback).

**Comment by ethanhs** [comment]
Ah, Windows doesn't allow less than or greater than in file paths. I am surprised that the tests didn't fail once there was an attempt to create the file however...

**Comment by gvanrossum** [comment]
> I am surprised that the tests didn't fail once there was an attempt to create the file however...

That's why I made the changes to testcmdline.py. It runs mypy in a subprocess, and the subprocess was printing a traceback. But there was logic in testcmdline.py that ignored all output (and the exit status) **if at least one output file was expected** (i.e. one or more `[outfile blah]` sections). It then complained it couldn't find the expected output file, but it didn't show the traceback.

I fixed it so it will now fail loudly and report the traceback.

**Pull request merged**